Update explainer.md (#7)
WICG/PEPC: explainer.md

However these measures have limited effect
[as indicated by metrics](#user-agent-abuse-mitigations).

There are four main challenges with the status quo:

> 1.  **Insufficiency of existing mitigations** The present day permissions spam

Reviewer: ```suggestion
1.  **Insufficiency of existing mitigations**: The present day permissions spam
```

Reviewer 2: Accepted

    and abuse mitigation approach has an architectural upper bound on user
    protection because the model relies on the website to choose when to trigger
    the permission request prompt rather than capturing a reliable signal of

>     user intent. Requiring a user gesture for the Permission API (or similar)
>     does not solve this problem as there are many ways of tricking a user into
>     providing such a gesture.

Reviewer: Note that this is not a function of the Permission API, but it is relate to the web's permission model (which is owned by the permission spec... it might be more accurate to say:  

```suggestion
    user intent. Requiring a user gesture to [request permission to use a powerful feature](https://www.w3.org/TR/permissions/#dfn-request-permission-to-use) (or similar)
    does not solve this problem as there are many ways of tricking a user into
    providing a so called "[activation triggering input event](https://html.spec.whatwg.org/#activation-triggering-input-event)" (i.e., a user gesture) .
```

Reviewer 2: Accepted

1.  **Context**: Ideally, a site's developer will request access as part of a
    contextual flow that helps users understand what's being asked for and why,
    enabling quick and confident responses. Often, however, permission requests
    are correlated poorly with user expectations, up to and including prompts
    that can come out of nowhere (see example 1). This places a burden on user

>     agents' presentation of the request. The user agent has no semantic

Reviewer: Maybe not worth mentioning... but the only thing UAs really have is "time to interaction"... why notifications get suppressed if they try to pop up too quickly (and without a user gesture, as can be the case with the Notificaitons API and Geolocation API).

Reviewer 2: Will add in a follow-up PR.

    understanding of events taking place in the content area prior to the
    permission request. User agents could make better decisions and provide
    better prompts if they could make well-founded assumptions about the nature
    of the user's interaction in the content area, and the user's intent.

    ![](images/image1.png) \

>     *Example 1. A notification permission prompt on a news site (contents

Reviewer: This example seems contrived... for Notifications, the API doesn't require a user gesture to request permissions. That's a flaw in the Notifications API (which we should fix collectively as user agents), but the notification prompt is not resulting from the user clicking - the site can show the permission prompt at any time.

Reviewer 2: This is a real-world scenario that I've personally experienced several times. The notifications API might not require a user gesture, but for some reason it seems that some sites will request the notification permission as soon as a click happens in the page.

    blurred), shown after the user has clicked on the empty area next to the
    article content. The user finds this prompt interruptive as they had no
    interest in subscribing to notifications, and they will likely struggle to
    understand why the prompt was shown to begin with.*

1.  **Location**: In the ideal case above, users will interact with something on
    a site that triggers a prompt. In less ideal cases, the user might not have
    interacted with anything at all, or they may have interacted with an element
    that was unrelated to the request. Given this uncertainty, user agents rely
    on common placement of the permission prompt, usually in the top-left of the
    page. Even in the best case, this has the unfortunate effect of shifting the
    point to which users need to pay attention from the thing they clicked on to
    some distant part of the user agent's UI (see example 2). User agents could

>     make better decisions and provide better prompts if they could make

Reviewer: General feedback... There is definitely scope to do something different, specially on desktop browsers... but it's feels a bit presumptuous to assume the general UX model will change. Increasingly (on Apple's platforms at least), the permissions requests experience (UX and UI) are all governed by the underlying OS. There are some exceptions, like with getUserMedia(), but even there, the permission to access the camera and microphone are still OS level prompts I believe.

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

    well-founded assumptions about the nature of the user's interaction in the
    content area, and the user's current area of focus. In effect, we think
    there is a benefit to semantic markup for permissions.

    ![](images/image2.png) \
    *Example 2. An example where the permission prompt is far away from the
    user's current area of focus. The permission prompt was triggered because
    the user has just clicked on the crosshair icon in the bottom right, but the

>     prompt is easy to miss since it's on the opposite side of the page.*

Reviewer: Again, this is not a universal issue. For example, on iOS this wouldn't be an issue because the prompt is modal. 

I think we have to be careful to not design features to overcome UX limitations that could be addressed in other ways. For instance, there is nothing stopping browsers from not doing the door hanger model for permissions.  
  
![Screenshot 2023-12-20 at 11 45 38](https://github.com/WICG/PEPC/assets/870154/0e2e05ab-187b-4b77-a716-ed1418754659)

In fact, notice that Safari doesn't do the door hanger for the Notifications API. It does the modal prompt on both macOs and iOS.

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

1.  **Regret**: Given the challenges of permission annoyance and abuse, it is
    reasonable for user agents to suppress a site's future requests for the same
    capability when the first request is blocked. That said, our research shows
    that users can and do change their minds for good reasons. When they change
    their mind, the site can no longer offer an interface in web content and the
    user must search for the appropriate user agent surface. Our research shows

>     that users often fail when trying to do so (see example 3). In these cases,

Reviewer: this is true... but again, this is a browser UI flaw that could be addressed at the browser level. 

Chrome itself seems to have attempted to address a bit of this, but bringing site settings directly into the address bar:  

![Screenshot 2023-12-20 at 11 50 31](https://github.com/WICG/PEPC/assets/870154/662ee4c9-f4ef-4dfb-8ef1-9232eccefa11)

There is not reason, why the permission toggles for a site couldn't be shown in that drop down. I guess what I'm trying to say is that we might not have exhausted all UX affordances yet at the browser level. 

I'm worried that we are shifting the problem from browser developers to could end up putting the permissions button in weird and awkward places too... so users would also need to hunt around a site for them and it could be just as bad, if not worse than what browsers do.

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

    the user agent's desire to protect the user backfires, and makes the user's
    experience worse as the site will not work as the user wants. User agents
    can help users recover from a permission regret state if they can make
    well-founded assumptions about the nature of the user's intent and
    interaction with web content.

    ![](images/image3.png) \
    *Example 3. An example where the user previously blocked camera and
    microphone access, but has now just expressed a strong intention to
    re-enable them by clicking the unmute buttons. Because the user agent has no
    insight into this interaction in the content area, it is compelled to
    respect the user's previous decision. Especially in a stressful scenario

>     such as an important presentation, users will struggle to navigate the

Reviewer: Right, but even here, this is a design flaw in the browser's UI. Making it easier to access the site settings may elevate some of this.

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

will be a drop-in replacement that is straightforward to adopt and easy to
polyfill on browsers which do not support the PEPC. Here are some real-life
examples:

![](images/image4.png) \
*Example 4: A video-conferencing site. Clicking on the "Enable camera" button
triggers a camera permission request.*

![](images/image5.png) \
*Example 5: A search site. Clicking on "Use precise location" triggers a
geolocation permission request.*

![](images/image6.png) \
*Example 6: A messaging site, clicking on the "Enable Desktop Notifications"
button triggers a push notifications permission request.*

We believe that enshrining such a user-initiated approach in standards can
contribute to consistently better permission request flows across the web. This
is because the permission element offers the following compelling advantages to
users and developers alike:

-   It is **non-interruptive**: it is static, small, and contained in the
    content area on the same z-level.
-   It is **discoverable**: it can be placed by the developer within the user's
    focus of attention; with the locality making it easier to find and more
    convenient to interact with.

> -   It provides more **contextual information**: it has a visual manifestation

Reviewer: There might be an accessibility case to be made here too (i.e., making it quick and convenient for assistive technologies to find these buttons on a page).

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

    as opposed to being a procedural API, requiring developers to think about
    integrating it into the user journey at UX design time, as opposed to being
    left as an afterthought during implementation, resulting in knock-on effects

>     relating to clearer context.

Reviewer: These are all achievable with `<button>` or any relevant HTML input (i.e., I would argue that, apart from maybe getting the free text label, there is limited advantage listed in these first 3 points over existing solutions). A developer would still need to do to support browsers that don't implement this feature:

```HTML
<permission type="geolocation"><button onclick="doGeo()">Use precise location</button></button>
```

Consider that image 5 has required significant styling for integration into the page's content:  
![Screenshot 2023-12-20 at 13 35 38](https://github.com/WICG/PEPC/assets/870154/e2d976de-7729-4039-af44-b50df964d385)

So it's a bit unconvincing that it would be an "afterthought during implementation, resulting in knock-on effects". A developer still needs to deal with the exact same situation in all cases.

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

> -   It allows users to **revert** a previous "deny" decision if they have

Reviewer: I think this is the core of the problem: how can users easily get to site settings? This solutions doesn't solve the general case where we have permissions being granted in other ways (or through "legacy" means... e.g., `Notifications.requestPermission()`).

Native apps (on iOS at least) solve for this by allowing users to go directly from the application to the appropriate permissions sheet. Irrespective of `<permission>`, we may still require something similar for web apps because otherwise everyone would need to rewrite their sites to use `<permission>` as a recovery mechanism - or not all APIs might be supported via `<permission>` (though if we did get agreement to support it, we certainly could make that the case).

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

control permissions on various UI surfaces that are entirely separate from the
site's rendering area). Therefore events specific to the PEPC will only deal
with the user's actions on the Permission UI, and specifically with the user
closing it either by dismissing it or by taking some other action on it that
causes it to close (e.g. they accept it). This allows sites to respond to this
event by providing more context to potentially help the user make a decision.
These two events will be added to
[GlobalEventHandlers](https://html.spec.whatwg.org/#globaleventhandlers) and can
only target `permission` HTML elements. They do not bubble and are not
cancelable.

> -   `ondismiss` - raised when the permission UI triggered by the PEPC has been

Reviewer: ```suggestion
-   `ondismiss` - fired when the permission UI triggered by the PEPC has been
```

Reviewer 2: To be fixed in follow-up PR

    dismissed by the user (for example via clicking the 'x' button or clicking
    outside the prompt)

> -   `onresolve` - raised when the permission UI triggered by the PEPC has been

Reviewer: ```suggestion
-   `onresolve` - fired when the permission UI triggered by the PEPC has been
```

Reviewer: "resolve" is very much "promise terminology"... might be good to find an alternative even name.

Reviewer 2: To be fixed in follow-up PR

abuse. Therefore the user agent should consider the following potentially
malicious tactics and mitigate them:

-   The site could trick the user by choosing some misleading text (e.g. "Click
    here to proceed"). Therefore the text on the PEPC should not be able to be
    set by the site, instead the user agent should make sure to set it to
    something comprehensive (e.g. "Share location" for a geolocation PEPC). \

>     **Open question:**should there be a mechanism that allows the site to pick

Reviewer: I think this is where other solutions have broken down (I hate to say it, but the iOS one)... it's so customizable to the point that it becomes meaningless. 

And for Geolocation, for instance, sometimes you really just want a button that's as simple as: 
![Screenshot 2023-12-20 at 14 02 39](https://github.com/WICG/PEPC/assets/870154/a5766d64-24d9-44e6-b024-a6553517dcd0)

Providing native text is also super challenging, because each browser might decide to provide different text... otherwise we would need to standardize every text variation for every possible language for every possible option. 

That's going to be super challenging because if one browser chooses one word, another chooses another, and that causes a line break, then the page's design will break.

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

    one of several flavors of text (example: "Share location" vs "Use
    location")?

> -   The style of the PEPC can be set to obscure the purpose (e.g. setting the

Reviewer: There may be good reason to do this... like, you want the entire page to fade in or change colors over time.

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

    same text color and button color would make the text unreadable). Therefore
    the style should be verified, validated and overridden by the user agent as
    needed. More details in the
    [Locking the PEPC style](#locking-the-pepc-style) section

> -   The PEPC might be partially covered (to hide the text) with another HTML

Reviewer: These kinda of heuristics can quickly become an issue, as developers will constantly be trying to work around them... we end up fighting content producers at the detriment of users and their experience.

    element. Therefore the user agent should verify that the PEPC has been
    visible already for some short time (e.g. 500ms or so) before it's clicked.
    User agents that implement the
    [IntersectionObserverV2](https://github.com/w3c/IntersectionObserver/blob/v2/explainer.md)
    API can make use of it internally.

> -   The site might try to obtain a click on the PEPC by moving it where the user

Reviewer: As above... if we are having to fight developers... then that's not great. This is a 🚩 in the design of this proposal. 

That's not to say that what we have right now is adequate, but having to be so combative with mitigations calls into question the proposal.

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8

There are 3 main possible approaches to consider, if the integrity of the PEPC
click is not assured:

> -   The click triggers the legacy permission flow (as if it was triggered by the

Reviewer: This is also the bit we (WebKit) are finding troubling, tbh. We went down this path previously with `permission.request()` and ultimately rejected the idea because it was duplicative of what's already implemented via JS.

The approach we prefer is to look at if we can fix things at the API level: 

 * https://github.com/w3c/geolocation-api/issues/48
 * https://github.com/whatwg/notifications/issues/108

Reviewer 2: Extracted into https://github.com/WICG/PEPC/issues/8
